tiledb/api/src/datatype.rs

```diff
@@ -0,0 +1,161 @@
+#[macro_export]
+macro_rules! fn_typed {
+    ($func:ident, $datatype:expr$(, $arg:expr)* => $then:expr) => {{
```

Comment:
Trying to mimic
```
template <class Fn, class... Args>
inline auto apply_with_type(Fn&& f, Datatype type, Args&&... args)
```
without variadic generics.

This first rule supports `fn_typed!(generic_function, my_datatype, arg1, arg2, ..., argN => /* do stuff */)`.  The only arguable wart is the `let generic_function = generic_function(arg1, ..., argN);` name shadowing but I don't really mind it much.